File: ngs/ngs-sdk/adapter/ReadItf.cpp
/*===========================================================================
*
*                            PUBLIC DOMAIN NOTICE
*               National Center for Biotechnology Information
*
*  This software/database is a "United States Government Work" under the
*  terms of the United States Copyright Act.  It was written as part of
*  the author's official duties as a United States Government employee and
*  thus cannot be copyrighted.  This software/database is freely available
*  to the public for use. The National Library of Medicine and the U.S.
*  Government have not placed any restriction on its use or reproduction.
*
*  Although all reasonable efforts have been taken to ensure the accuracy
*  and reliability of the software and data, the NLM and the U.S.
*  Government do not and cannot warrant the performance or results that
*  may be obtained by using this software or data. The NLM and the U.S.
*  Government disclaim all warranties, express or implied, including
*  warranties of performance, merchantability or fitness for any particular
*  purpose.
*
*  Please cite the author in any work or product based on this material.
*
* ===========================================================================
*
*/

#include <ngs/adapter/ReadItf.hpp>
#include <ngs/adapter/StringItf.hpp>
#include <ngs/adapter/ErrorMsg.hpp>

#include "ErrBlock.hpp"

namespace ngs_adapt
{

    /*----------------------------------------------------------------------
     * ReadItf
     */

    ReadItf :: ReadItf ()
        : FragmentItf ( & ivt . dad )
    {
    }

    NGS_String_v1 * CC ReadItf :: get_id ( const NGS_Read_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const ReadItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getReadId ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    uint32_t CC ReadItf :: get_num_frags ( const NGS_Read_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const ReadItf * self = Self ( iself );
        try
        {
            return self -> getNumFragments ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    uint32_t CC ReadItf :: get_category ( const NGS_Read_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const ReadItf * self = Self ( iself );
        try
        {
            return self -> getReadCategory ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC ReadItf :: get_read_group ( const NGS_Read_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const ReadItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getReadGroup ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC ReadItf :: get_name ( const NGS_Read_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const ReadItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getReadName ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC ReadItf :: get_bases ( const NGS_Read_v1 * iself, NGS_ErrBlock_v1 * err, uint64_t offset, uint64_t length )
    {
        const ReadItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getReadBases ( offset, length );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC ReadItf :: get_quals ( const NGS_Read_v1 * iself, NGS_ErrBlock_v1 * err, uint64_t offset, uint64_t length )
    {
        const ReadItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getReadQualities ( offset, length );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    bool CC ReadItf :: next ( NGS_Read_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        ReadItf * self = Self ( iself );
        try
        {
            return self -> nextRead ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return false;
    }

    NGS_Read_v1_vt ReadItf :: ivt =
    {
        {
            "ngs_adapt::ReadItf",
            "NGS_Read_v1",
            0,
            & FragmentItf :: ivt . dad
        },

        get_id,
        get_num_frags,
        get_category,
        get_read_group,
        get_name,
        get_bases,
        get_quals,
        next
    };

} // namespace ngs_adapt

File: ngs/ngs-java/ngs/itf/ReadItf.java
/*===========================================================================
*
*                            PUBLIC DOMAIN NOTICE
*               National Center for Biotechnology Information
*
*  This software/database is a "United States Government Work" under the
*  terms of the United States Copyright Act.  It was written as part of
*  the author's official duties as a United States Government employee and
*  thus cannot be copyrighted.  This software/database is freely available
*  to the public for use. The National Library of Medicine and the U.S.
*  Government have not placed any restriction on its use or reproduction.
*
*  Although all reasonable efforts have been taken to ensure the accuracy
*  and reliability of the software and data, the NLM and the U.S.
*  Government do not and cannot warrant the performance or results that
*  may be obtained by using this software or data. The NLM and the U.S.
*  Government disclaim all warranties, express or implied, including
*  warranties of performance, merchantability or fitness for any particular
*  purpose.
*
*  Please cite the author in any work or product based on this material.
*
* ===========================================================================
*
*/

package ngs.itf;

import ngs.ErrorMsg;
import ngs.Read;
import ngs.Fragment;
import ngs.FragmentIterator;


/*==========================================================================
 * ReadItf
 *  represents an NGS machine read
 *  having some number of biological Fragments
 */
class ReadItf
    extends Refcount
    implements Read
{

    /**********************
     * Fragment Interface *
     **********************/

    /* getFragmentId
     */
    public String getFragmentId ()
        throws ErrorMsg
    {
        return this . GetFragmentId ( self );
    }

    /* getFragmentBases
     *  return sequence bases
     *  "offset" is zero-based
     *  "length" must be >= 0
     */
    public String getFragmentBases ()
        throws ErrorMsg
    {
        return this . GetFragmentBases ( self, 0, -1 );
    }

    public String getFragmentBases ( long offset )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );

        return this . GetFragmentBases ( self, offset, -1 );
    }

    public String getFragmentBases ( long offset, long length )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );
        if ( length < 0 )
            throw new IndexOutOfBoundsException ( "length " + length + " is negative");

        return this . GetFragmentBases ( self, offset, length );
    }


    /* getFragmentQualities
     *  return phred quality values
     *  using ASCII offset of 33
     *  "offset" is zero-based
     *  "length" must be >= 0
     */
    public String getFragmentQualities ()
        throws ErrorMsg
    {
        return this . GetFragmentQualities ( self, 0, -1 );
    }

    public String getFragmentQualities ( long offset )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );

        return this . getFragmentQualities ( offset, -1 );
    }

    public String getFragmentQualities ( long offset, long length )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );
        if ( length < 0 )
            throw new IndexOutOfBoundsException ( "length " + length + " is negative");

        return this . GetFragmentQualities ( self, offset, length );
    }

    public boolean isPaired ()
        throws ErrorMsg
    {
        return this . IsPaired ( self );
    }
    
    public boolean isAligned ()
        throws ErrorMsg
    {
        return this . IsAligned ( self );
    }

    /******************************
     * FragmentIterator Interface *
     ******************************/

    /* nextFragment
     *  advance to next Fragment
     *  returns false if no more Fragments are available.
     *  throws exception if more Fragments should be available,
     *  but could not be accessed.
     */
    public boolean nextFragment ()
        throws ErrorMsg
    {
        return this . NextFragment ( self );
    }

    /******************
     * Read Interface *
     ******************/

    /* getReadId
     */
    public String getReadId ()
        throws ErrorMsg
    {
        return this . GetReadId ( self );
    }

    /* getNumFragments
     *  returns the number of biological Fragments contained in the read
     */
    public int getNumFragments ()
        throws ErrorMsg
    {
        return this . GetNumFragments ( self );
    }
    
    public boolean fragmentIsAligned ( int fragIdx )
        throws ErrorMsg
    {
        return this . FragmentIsAligned ( self, fragIdx );
    }


    /*----------------------------------------------------------------------
     * read details
     */

    /* getReadCategory
     */
    public int getReadCategory ()
        throws ErrorMsg
    {
        return this . GetReadCategory ( self );
    }

    /* getReadGroup
     */
    public String getReadGroup ()
        throws ErrorMsg
    {
        return this . GetReadGroup ( self );
    }

    /* getReadName
     */
    public String getReadName ()
        throws ErrorMsg
    {
        return this . GetReadName ( self );
    }


    /* getReadBases
     *  return sequence bases
     *  "offset" is zero-based
     *  "length" must be >= 0
     */
    public String getReadBases ()
        throws ErrorMsg
    {
        return this . GetReadBases ( self, 0, -1 );
    }

    public String getReadBases ( long offset )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );

        return this . GetReadBases ( self, offset, -1 );
    }

    public String getReadBases ( long offset, long length )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );
        if ( length < 0 )
            throw new IndexOutOfBoundsException ( "length " + length + " is negative");

        return this . GetReadBases ( self, offset, length );
    }


    /* getReadQualities
     *  return phred quality values
     *  using ASCII offset of 33
     *  "offset" is zero-based
     *  "length" must be >= 0
     */
    public String getReadQualities ()
        throws ErrorMsg
    {
        return this . GetReadQualities ( self, 0, -1 );
    }

    public String getReadQualities ( long offset )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );

        return this . getReadQualities ( offset, -1 );
    }

    public String getReadQualities ( long offset, long length )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );
        if ( length < 0 )
            throw new IndexOutOfBoundsException ( "length " + length + " is negative");

        return this . GetReadQualities ( self, offset, length );
    }


    /***************************
     * ReadItf Implementation *
     ***************************/


    // constructors
    ReadItf ( long ref )
    {
        super ( ref );
    }

    ReadItf ( Read obj )
        throws ErrorMsg
    {
        super ( 0 );
        try
        {
            ReadItf ref = ( ReadItf ) obj;
            this . self = ref . duplicate ();
        }
        catch ( Exception x )
        {
            throw new ErrorMsg ( x . toString () );
        }
    }

    // native interface
    private native String GetFragmentId ( long self )
        throws ErrorMsg;
    private native String GetFragmentBases ( long self, long offset, long length )
        throws ErrorMsg;
    private native String GetFragmentQualities ( long self, long offset, long length )
        throws ErrorMsg;
    private native boolean IsPaired ( long self )
        throws ErrorMsg;
    private native boolean IsAligned ( long self )
        throws ErrorMsg;
    private native boolean NextFragment ( long self )
        throws ErrorMsg;
    private native String GetReadId ( long self )
        throws ErrorMsg;
    private native int GetNumFragments ( long self )
        throws ErrorMsg;
    private native boolean FragmentIsAligned ( long self, int fragIdx )
        throws ErrorMsg;
    private native int GetReadCategory ( long self )
        throws ErrorMsg;
    private native String GetReadGroup ( long self )
        throws ErrorMsg;
    private native String GetReadName ( long self )
        throws ErrorMsg;
    private native String GetReadBases ( long self, long offset, long length )
        throws ErrorMsg;
    private native String GetReadQualities ( long self, long offset, long length )
        throws ErrorMsg;
}
